keps/sig-api-machinery/3617-fine-grained-authz/README.md

```diff
@@ -0,0 +1,1162 @@
+   requires a large number of authz requests.
+
+When fine-grained field permissions need to be checked, they will be checked
+in the order of most general to most specific. When we describe the
+representation in the schema it will be clear what that means. The reason for
+this is that multiple fields can be covered by a more general permission,
+greatly reducing the number of checks needed in the worst case (as long as the
+system administrator has made use of the general permission).
+
+Fields with an "uncovered" permission don't get the above hierarchical
+treatment; you must have that exact permission.
+
+### Notes/Constraints/Caveats (Optional)
+
+This is a complex design. See the alternatives for why we propose it anyway.
+
+### Risks and Mitigations
+
+This design avoids the "grant an overly broad permission and then restrict it
+later" pattern, which has undesirable failure / misconfiguration patterns.
+
+This design is written to have no impact on existing cluster users for existing
+fields.
+
+A risk is that it would be possible to configure permissions which result in a
+large number of authz checks. Specifically CRD authors could craft a CRD with
+many checks required to do anything.
+
+A runtime risk is an actor with the "granular" permission could make requests
+changing many fields it doesn't have permission for, with the aim of driving
+many authz checks to overload the authz system. To mitigate this we (a) have the
+"granular" permission at all, to reduce the number of actors that could attempt
+this, and (b) we will not check all special permissions for a request once we
+know at least one has failed.
+
+## Design Details
+
+### Permission Description System
+
+Now that we have described all requirements and given the overall outline, we
+can finally state exactly how the mapping from field to permission works and how
+it is specified. Additionally, we will describe the parameterization systems. We
+will give a series of examples in {goal, built-in specification, CRD
+specification, resulting verb} tuplets.
+
+##### Everything in spec
+
+On types:
+
+```go
+// +permission-verb:"specification"
+type FooSpec struct {
+```
+
+Or on fields:
+
+```go
+  ...
+  // +permission-verb:"specification"
+  Spec FooSpec
+  ...
+```
+
+On CRDs schemas:
+
+```
+"x-kubernetes-permission-verb": "specification"
+```
+
+The authz system will see the verb "granular:specfication".
+
+This illustrates that the granular permissions can be made hirearchical;
+granting "granular" and "granular:specification" to some actor permits it to
+modify anything in .spec. API Server will check the permissions from most
+general to most specific, so that a single general permission can let it avoid
+checking many specific permissions.
+
+#### Pod's nodeName
+
+Given where this appears in Pod, this permission will only be checked for pods
+if the agent doesn't have the "specification" permission.
```

Comment:
If they have PATCH/PUT it is also satisifed, right?  Because this is a convered permission?